### ISSUE TITLE
Non recursive GADT-style type definitions

### ISSUE THREAD

**Issue by SkySkimmer** [Andromedans/andromeda#289]
Non recursive definitions in GADT style have a problem: the type being defined still needs to be bound in the return type of its constructors.
See http://caml.inria.fr/mantis/view.php?id=6934 especially the first comment:
> I'd be in favor of just updating the error message to say that you can't use `nonrec` when defining a GADT. I don't like the idea of allowing `type nonrec t = A : t -> t` where the last two `t`s means two different things.

That's not a concern if we don't allow type shadowing, but if when we add modules we still get a form of shadowing, so it doesn't seem like a good answer:
```
(* outer module *)
type t = foo
module Bar = struct
  type t = T : t (* the outer one *) -> t (* the inner one *)
end
```

The main reason for avoiding `of` is that we don't want to overload `*` to get >1 arity constructors.
Maybe we can use `and` instead of `*`?
```
type pair a b =
  | Pair of a and b
  end
and triple a b c =
  | Triple of a and b and c
  end
```

It may be a bit weird to have to use `type rec` for GADTs though:
```
type rec 'a int_or_string =
  | Int : int -> int int_or_string
  | String : string -> string int_or_string
```

**Comment by andrejbauer** [comment]
I wouldn't worry about the concrete syntax too much at this point, as long as we see how to translate to Ocaml without too many complications. We can use the GADT syntax but treat type definitions as non-GADT (surely we're not going to implement GADTs).

**Comment by SkySkimmer** [comment]
Closed by #291: removed the GADT style syntax.
